test/new-e2e/pkg/utils/e2e/client/agent_client.go

```diff
@@ -169,3 +171,97 @@ func curlCommand(endpoint string, authToken string) string {
 		endpoint,
 	)
 }
+
+func waitForReadyTimeout(t *testing.T, host *Host, commandRunner *agentCommandRunner, timeout time.Duration) error {
+	err := commandRunner.waitForReadyTimeout(timeout)
+
+	if err != nil {
+		// Propagate the original error if we have another error here
+		localErr := generateAndDownloadFlare(t, commandRunner, host)
+
+		if localErr != nil {
+			t.Errorf("Could not generate and get a flare: %v", localErr)
+		}
+	}
+
+	return err
+}
+
+func generateAndDownloadFlare(t *testing.T, commandRunner *agentCommandRunner, host *Host) error {
+	profile := runner.GetProfile()
+	outputDir, err := profile.GetOutputDir()
+	flareFound := false
+
+	if err != nil {
+		return fmt.Errorf("could not get output directory: %v", err)
+	}
+
+	_, err = commandRunner.FlareWithError(agentclient.WithArgs([]string{"--email", "e2e@test.com", "--send", "--local"}))
+	if err != nil {
+		t.Errorf("Error while generating the flare: %v.", err)
+		// Do not return now, the flare may be generated locally but was not uploaded because there's no fake intake
+	}
+
+	flareRegex, err := regexp.Compile(`datadog-agent-.*\.zip`)
+	if err != nil {
+		return fmt.Errorf("could not compile regex: %v", err)
+	}
+
+	tmpFolder, err := host.GetTmpFolder()
+	if err != nil {
+		return fmt.Errorf("could not get tmp folder: %v", err)
+	}
+
+	entries, err := host.ReadDir(tmpFolder)
+	if err != nil {
+		return fmt.Errorf("could not read directory: %v", err)
+	}
+
+	for _, entry := range entries {
+		if flareRegex.MatchString(entry.Name()) {
+			t.Logf("Found flare file: %s", entry.Name())
+
+			if host.osFamily != osComp.WindowsFamily {
+				_, err = host.Execute(fmt.Sprintf("sudo chmod 744 %s/%s", tmpFolder, entry.Name()))
+				if err != nil {
+					return fmt.Errorf("could not update permission of flare file %s/%s : %v", tmpFolder, entry.Name(), err)
+				}
+			}
+
+			t.Logf("Downloading flare file in: %s", outputDir)
+			err = host.GetFile(fmt.Sprintf("%s/%s", tmpFolder, entry.Name()), fmt.Sprintf("%s/%s", outputDir, entry.Name()))
+
+			if err != nil {
+				return fmt.Errorf("could not download flare file from %s/%s : %v", tmpFolder, entry.Name(), err)
+			}
+
+			flareFound = true
+		}
+	}
+
+	if !flareFound {
+		t.Errorf("Could not find a flare. Retrieving logs directly instead...")
+
+		logsFolder, err := host.GetLogsFolder()
+		if err != nil {
+			return fmt.Errorf("could not get logs folder: %v", err)
+		}
+
+		entries, err = host.ReadDir(logsFolder)
+
+		if err != nil {
+			return fmt.Errorf("could not read directory: %v", err)
+		}
+
+		for _, entry := range entries {
+			t.Logf("Found log file: %s. Downloading file in: %s", entry.Name(), outputDir)
+
+			err = host.GetFile(fmt.Sprintf("%s/%s", logsFolder, entry.Name()), fmt.Sprintf("%s/%s", outputDir, entry.Name()))
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**❓ question**
Should we invert the slash when downloading from a Windows host?

Reply:
Yes we should but there's no need to do it explicitly because this is done in the [GetFile](https://github.com/DataDog/datadog-agent/blob/d3c8b8d11d579cf03a786042cbd5249d06ab91f1/test/new-e2e/pkg/utils/e2e/client/host.go#L175) function